File: backend/app/schemas/event.py
"""Schemas for the security events API (Phase 1 Step 4.4).

An "event" is the SOC-facing view of an AlertGroup: one aggregated
security event with N evidence alerts.
"""
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class EventListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    category: str
    severity: str
    status: str
    alert_count: int
    first_seen: datetime
    last_seen: datetime
    # Current risk assessment (Step 5.4) — None for legacy events without
    # a risk record; populated by the API layer from group.risk.
    risk_score: int | None = None
    risk_level: str | None = None


class EventListResponse(BaseModel):
    total: int
    page: int
    size: int
    items: list[EventListItem]


class EventAlertItem(BaseModel):
    """One evidence alert belonging to an event."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    source: str
    event_type: str
    severity: str
    title: str | None = None
    host_name: str | None = None
    source_ip: str | None = None
    created_at: datetime


class EventInfo(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    fingerprint: str
    title: str
    category: str
    severity: str
    status: str
    alert_count: int
    first_seen: datetime
    last_seen: datetime
    created_at: datetime
    updated_at: datetime


class RiskFactorItem(BaseModel):
    """One explainable contribution to the risk score ({name, score, reason})."""

    name: str
    score: int
    reason: str


class EventRiskDetail(BaseModel):
    """Full risk assessment of an event, including the factor breakdown."""

    model_config = ConfigDict(from_attributes=True)

    score: int
    level: str
    factors: list[RiskFactorItem]
    updated_at: datetime


class EventDetailResponse(BaseModel):
    """GET /events/{id}: event summary + evidence alerts + risk detail."""

    event: EventInfo
    alerts: list[EventAlertItem]
    risk: EventRiskDetail | None = None

File: frontend/src/types/event.ts
/**
 * TypeScript mirror of the Events API schemas (Phase 1 Step 4.4 / 5.4).
 * Backend serialises UUID/datetime as strings; keep them as strings here.
 */

export type EventSeverity = 'low' | 'medium' | 'high' | 'critical'
export type RiskLevel = 'low' | 'medium' | 'high' | 'critical'

export interface EventListItem {
  id: string
  title: string
  category: string
  severity: string
  status: string
  alert_count: number
  first_seen: string
  last_seen: string
  /** Current risk snapshot; null for legacy events without a risk record. */
  risk_score: number | null
  risk_level: string | null
}

export interface EventListResponse {
  total: number
  page: number
  size: number
  items: EventListItem[]
}

export interface EventAlertItem {
  id: string
  source: string
  event_type: string
  severity: string
  title: string | null
  host_name: string | null
  source_ip: string | null
  created_at: string
}

export interface EventInfo {
  id: string
  fingerprint: string
  title: string
  category: string
  severity: string
  status: string
  alert_count: number
  first_seen: string
  last_seen: string
  created_at: string
  updated_at: string
}

export interface RiskFactorItem {
  name: string
  score: number
  reason: string
}

export interface EventRiskDetail {
  score: number
  level: string
  factors: RiskFactorItem[]
  updated_at: string
}

export interface EventDetailResponse {
  event: EventInfo
  alerts: EventAlertItem[]
  risk: EventRiskDetail | null
}
